src/main/scala/org/ergoplatform/nodeView/state/UtxoState.scala

```diff
@@ -101,37 +99,61 @@ class UtxoState(override val persistentProver: PersistentBatchAVLProver[Digest32
         val inRoot = rootHash
 
         val stateTry = stateContext.appendFullBlock(fb).flatMap { newStateContext =>
-          val tm0 = System.currentTimeMillis()
           val txsTry = applyTransactions(fb.blockTransactions.txs, fb.header.stateRoot, newStateContext)
-          val tm = System.currentTimeMillis()
-          log.debug(s"Transactions at height $height checked in ${tm-tm0} ms.")
 
           txsTry.map { _: Unit =>
             val emissionBox = extractEmissionBox(fb)
             val meta = metadata(idToVersion(fb.id), fb.header.stateRoot, emissionBox, newStateContext)
 
-            val tp0 = System.currentTimeMillis()
-            val proofBytes = persistentProver.generateProofAndUpdateStorage(meta)
-            val tp = System.currentTimeMillis()
-            log.debug(s"Utxo storage at height $height updated in ${tp-tp0} ms.")
+            var proofBytes = persistentProver.generateProofAndUpdateStorage(meta)
+
+            if (!store.get(scorex.core.idToBytes(fb.id)).exists(w => java.util.Arrays.equals(w, fb.header.stateRoot))) {
+              throw new Error("Storage kept roothash is not equal to the declared one")
+            }
+
+            if (!java.util.Arrays.equals(fb.header.stateRoot, persistentProver.digest)) {
+              throw new Error("Calculated stateRoot is not equal to the declared one")
+            }
+
+            var proofHash = ADProofs.proofDigest(proofBytes)
+
+            if (!java.util.Arrays.equals(fb.header.ADProofsRoot, proofHash)) {
 
-            val proofHash = ADProofs.proofDigest(proofBytes)
+              log.error("Calculated proofHash is not equal to the declared one, doing another attempt")
+
+              /**
+                * Proof generated was different from one announced.
+                *
+                * In most cases, announced proof is okay, and as proof is already checked, problem in some
+                * extra bytes added to the proof.
+                *
+                * Could be related to https://github.com/ergoplatform/ergo/issues/1614
+                *
+                * So the problem could appear on mining nodes only, and caused by
+                * proofsForTransactions() wasting the tree unexpectedly.
+                *
+                * We are trying to generate proof again now.
+                */
+
+              persistentProver.rollback(inRoot)
+                .ensuring(java.util.Arrays.equals(persistentProver.digest, inRoot))
+
+              val mods = ErgoState.stateChanges(fb.blockTransactions.txs).operations
+              mods.foreach(persistentProver.performOneOperation)
+
+              proofBytes = persistentProver.generateProofAndUpdateStorage(meta)
```

Comment:
Shouldn't meta be also recomputed again?

Reply:
No, as it is not dependent on proofs hash.